Refactor picklers/unpickler and add some changes to spores

jvican opened the pull request:
General refactoring of all the generation of picklers and unpicklers for spores.

It fixes #39 as a side effect.

jvican committed:
Port to scala pickling 0.11.x

* Still uses publishLocal since there is no official release of such
  version of scala-pickling.

jvican committed:
Refactor picklers and unpicklers

* Simplify picklers and unpicklers. Even though it can be further
  simplified into only one pickler/unpickler generation, right now
  we have two: one for `SporeWithEnv` and another one for `Spore`.

* Use the idea of `SelfDescribing` as the core concept of this
  refactoring. There is only one implicit unpickler for any kind of
  spore that will detect the className of the `Unpickler`, instantiate
  it and properly unpickle the message.

* Some tests are failing because of a bug in the binary format. Somehow,
  when it's supposed to only read `unpicklerClassName` it reads the
  spore `className` and it fails. If we use json instead, this error
  disappears. I'm currently working on a fix for it.

* Remove the `SelfDescribing` test since it's not required anymore.

* Follow DRY practices and simplify overall design.

* Make `isEffectivelyPrimitive` tail-recursive.

jvican commented:
5/20 tests are failing because of a bug in the binary format. Somehow, when it's supposed to only read `unpicklerClassName` it reads the spore `className` and it fails. This may have to do with the order in which the binary format writes them or reads them. If we use json instead, this error disappears. I'm currently working on a fix for it.

jvican commented:
Review by @heathermiller and @phaller.

jvican committed:
Simplify readers and writers

* Make readers and writers simpler by reusing a macro-based template.

* Add more documentation.

* Enable debugging (temporarily).

jvican committed:
Clean up some code and make it more robust

jvican committed:
Keep hygiene in macros

* Use paths and types in `Utils` to allow identifiers and variables from
  producing bad code in macro expansion.

jvican committed:
Add autoregister to picklers/unpicklers

* Use the new `AutoRegister` introduced in the last refactorings of
  scala pickling `0.11.x`.

jvican commented:
Build fails because spores-pickling `0.11.x` hasn't been released. Otherwise, it'll compile fine.
  
However, there are two failed tests due to an issue in the binary format. Apparently, it can't get a generator of tuple from `genPicklerMap`. I'll look into this problem in the next days.
  
The problem with binary format that I discussed with @phaller has disappeared because of some rearrangements in how we read and share readers between picklers/unpicklers.

jvican committed:
Explicit generation of picklers for captured types

* Don't get the picklers/unpicklers of captured types as implicits in
  the signature of the macros but look for them explicitly at the same
  place where the macro expansion is done. They are generated/looked up
  when the macro typechecks. This improves compilation times as it
  doesn't create a new pickler for `this.Captured` every time, which is
  the type used by the compiler instead of the actual dealiased type.

* Make sure we `dealias` always to get the right type instead of type
  aliases when using type refinements.

* Don't use `elideType` temporarily since we're going to register
  spore picklers in the `PicklerRegistry` to cache them and don't generate
  them by reflection.

* Update version of scala pickling with latest changes.

jvican committed:
Add registered picklers for basic spores

* Create registered picklers for spores for basic types (there's still
  no implementation for `SporeWithEnv`, only spores that don't capture
  anything from the environment).

* We cannot reuse the logic of the macros since that would force an
  execution of macros inside the same module that defines them.

* Create an independent object of `ReflectionUtils` that we will reuse
  also in the macros.

jvican committed:
Reuse `createInstance` in the macro

* `ReflectionUtils` contains now any reflection-related code and we
  reuse it in all the spore picklers.

jvican committed:
Add runtime tests for spore

* Add Runtime tests that show how a spore is supposed to be pickled when
  nothing is known about their type parameters or even when they are
  `Any`.

* Annotate that the `anonspore` name shouldn't be changed. We'll use it
  to tell from the class name if a certain class is a spore or not.

* Make sure that a spore has a non-null class name. Concrete picklers
  and unpicklers don't use it, but it's done for correctness.

jvican committed:
Disable debug option

* All tests pass with scala-pickling 0.11-M5

jvican committed:
Apply minor fixes to polish release

* Jump up to version 3.0 due to the huge amount of changes in spores and
spores-pickling.

* Make `ReflectionUtils` public because of visibility issues.

* Remove useless debug statement.

phaller commented on core/src/main/scala/scala/spores/package.scala:
```diff
@@ -44,7 +44,7 @@ package object spores {
 
   // TOGGLE DEBUGGING
-  private val isDebugEnabled = System.getProperty("spores.debug", "false").toBoolean
+  private val isDebugEnabled = System.getProperty("spores.debug", "true").toBoolean
```
Should probably not be true by default.

phaller commented on project/SporesBuild.scala:
```diff
@@ -9,8 +9,8 @@ object SporesBuild extends Build {
     organizationHomepage := Some(new URL("http://lamp.epfl.ch")),
-    version      := "0.2.2-SNAPSHOT",
-    scalaVersion := "2.11.6"
+    version      := "0.2.4-M1",
```
What happened to 0.2.3?

jvican replied:
This was a quick workaround to avoid overriding `0.2.3`, which i had already published and used locally. Changed.

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/PicklingBinary.scala:
```diff
@@ -12,72 +12,74 @@ import binary._
 @RunWith(classOf[JUnit4])
 class PicklingBinarySpec {
+
+  implicit val staticOnly = static.StaticOnly
```
I think an import would be clearer.

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/SelfDescribing.scala:
```diff
@@ -1,104 +0,0 @@
-import SporePickler._
-
-
-final case class SelfDescribing(unpicklerClassName: String, blob: Array[Byte]) {
```
I don't remember when this was supposed to go. Into the F-P repo?

jvican replied:
Yes :)

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/PicklingBinary.scala:
```diff
@@ -12,72 +12,74 @@ import binary._
 
-    val s = spore {
+    val maxSize = 20
+    val s: Spore2WithEnv[(String, Int), Emitter[String], Unit] {type Captured = (Int, Int)} = spore {
```
This looks to me like a different test, since the type of the spore is no longer inferred. How can we know for sure that the type of the spore did not change? Tests where the type of a spore is inferred are different from tests where the type is provided explicitly. Therefore, you should really make sure there remains a test which looks like the previous one.

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/Pickling.scala:
```diff
@@ -14,102 +14,75 @@ import SporePickler._
-    val v2 = "hello"
-    val s = spore {
+    val v2 = "hello2"
+    val s: SporeWithEnv[Int, String] {type Captured = (Int, String)}= spore {
```
Again, this is a different test from before, since the type of the spore is no longer inferred.

jvican commented:
This is again ready for review guys @heathermiller and @phaller. I'm trying to get it merged so that the function-passing CI works (locally, it has a coverage of 70% and is passing all the tests, but as the milestones of scala/pickling and spores are not published, it can't run).

phaller commented:
Sorry I dropped the ball here. @jvican could you let me know how I can help move this forward? Did you address all my previous comments?

jvican commented:
Yes, I did @phaller, have a look :smile:

jvican commented:
But let's not merge it now, I'm getting done the design that we discussed: being able to get implicits for types like `Spore {type Captured = (T, S)}`. So far, we support `Spore` and `SporeWithEnv {type Captured = (T, S)}`. It's almost ready, so when I have it i'll update this and then we merge it.
  
It would be better to have a review now, though!

phaller commented:
@jvican OK, cool. I'll try to do the review tomorrow then.

jvican committed:
Add Nothing as the captured type of simple spores

* Any simple spore will, by default, extends a spore with the following
  type refinement: `type Captured = Nothing`.

* Add test that confirm the change for `Spore`, `Spore2` and `Spore3`.

phaller commented on spores-pickling/src/main/scala/scala/spores/SporePicklers.scala:
```diff
@@ -0,0 +1,372 @@
+             * since such an `Unpickler` exists because we got it as an implicit */
+
+            val $capturedUnpickler = implicitly[$unpicklerType[$utpe]]
+            if($capturedUnpickler == null) println("Unpickler[" + ${utpe.toString} + "] is null")
```
We probably don't want to keep this `println` there, do we? Perhaps throwing an exception would be better.

phaller commented on spores-pickling/src/main/scala/scala/spores/SporePicklers.scala:
```diff
@@ -0,0 +1,372 @@
+  }
+
+  /** Generates an `Unpickler` that will be able to get the classname of another
+    * `Unpickler` and successfully instantiate it and delegate his responsability.
```
"delegate _its_ _responsibility_"

Last time I checked, unpicklers didn't have a gender. ;-)

phaller commented on spores-pickling/src/main/scala/scala/spores/SporePicklers.scala:
```diff
@@ -0,0 +1,372 @@
+
+  /******************************** Unpicklers ********************************/
+
+  /* These `Unpickler`s are meant to serialize both `Spore`s and `SporeWitEnv`s.
```
`SporeWitEnv` -> `SporeWithEnv`

phaller replied:
You mean _deserialize_ here, right? `FullU` is only an unpickler as far as I can see. Also, why not use the term "unpickle" instead of "deserialize", since that is more consistent with our usual terminology. ;-)

jvican replied:
Hahaha, good catch!

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/Issue15.scala:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    val d = s.pickle
+    System.out.println(d.value.toString)
```
Probably remove the `println`.

jvican replied:
You sure we want to remove it? By default it's not executed, only when tests fail. I find it very useful... when tests are green, the test report is clean, otherwise it's dirty.

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/Pickling.scala:
```diff
@@ -10,106 +18,85 @@ import scala.pickling._
     }
 
-    assert(res.value.toString.endsWith(""""captured": {
+    System.out.println(s"res1: ${res.value}")
```
`println`

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/Pickling.scala:
```diff
@@ -10,106 +18,85 @@ import scala.pickling._
+    val unpickler = SporePicklers.genSporeUnpickler[Int, String]
+    val up = unpickler.unpickle("", reader).asInstanceOf[Spore[Int, String]]
+    val res2 = up(5)
+    System.out.println(s"res2: ${res2.value}")
```
`println`

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/PicklingBinary.scala:
```diff
@@ -20,63 +27,66 @@ trait Emitter[T] {
 
   @Test
   def testSimpleSpore(): Unit = {
+
```
Btw, normally we don't have blank lines at the beginning of method bodies. So, I'm a bit reluctant to accept introducing this "convention" into existing code. I think new code should have a style similar to the existing code. This would also be consistent with Scala's code base. OK?

jvican replied:
Agreed!

phaller commented on spores-pickling/src/test/scala/scala/spores/run/pickling/PicklingBinary.scala:
```diff
@@ -94,23 +104,29 @@ class PicklingBinarySpec {
     s2(("hello, " -> 0), testEmitter)
     s2(("world!" -> 9), testEmitter)
     assert(testEmitter.builder.toString == "hello, world!")
+
```
Dito.

phaller commented:
@jvican I identified a few more issues, which are rather minor, but should still be addressed I think. Otherwise, it looks very good!

jvican committed:
Improve considerably spore types picklers

* Now, we're also able to find `Pickler` and `Unpickler` for `Spore`s
  that can be both `SporeWithEnv` or simple `Spore`s that don't capture
  anything from the environment. Those are expressed with type
  refinements, e.g. `Spore[T,S] {type Captured = C}`.

* Picklers and unpicklers for the above types need runtime instance
  checks because they can get both possibilities at runtime and it will
  typecheck.

* Make simple spores to extend `Spore[T,S] {type Captured = Nothing}`
  and allow picklers and unpicklers for them that are chosen at
  compile-time.

* Add more tests (a total amount of 23 tests check the correct
  functionality of this commit). More are gonna come in the future.

* Add some fixes to `RuntimeSporePicklers`.

* Remove type of variable to allow type inference take place and choose
  the right implicit.

* Ignore ensime in git.

jvican committed:
Address philipp's suggestions

 * Import staticOnly instead of putting it as a implicit.

jvican committed:
Fix scalac/JVM bug with `getSimpleName`

 * Replace `getSimpleName` by `getName` since the former one throws a
   `java.lang.InternalError: Malformed class name`.

jvican commented:
This is completely ready for review from now on! As soon as Heather cuts the scala pickling release, the CI should be green! I'll add more tests in the near future to make it super robust. So far, we've got 5 o 6 more than before (now it's around 23). And all of them are passing, yay!

jvican commented:
Oh, actually, I didn't see your review Philipp, I'm going to make your changes and push again, this one for the final version!

jvican committed:
Address last suggestions of Philipp

* Fix minor formatting issues and some typos.

phaller commented:
LGTM

phaller commented:
@jvican Could you rebase? Right now it doesn't merge cleanly.

jvican commented:
Rebased!

phaller commented:
@jvican Thanks!
